Add support for Cortex-M7 with single-precision FPU

aurel32 opened the pull request:
Some of the STM32F7 SoCs have a FPU limited to single precision. This patch series add support for this by adding a `CPU_HAS_FPU_DP` config option to enable double precision support. This option is not enable on the STM32F723 SoC.

Fixes: #8374

aurel32 committed:
boards: arm: remove CPU_HAS_FPU from defconfig

CPU_HAS_FPU is a hidden property from the SoC, it's not something
configurable by the end user, and thus should not be defined in the
defconfig file.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 commented:
recheck

ioannisg commented on arch/arm/Kconfig:
```diff
@@ -19,4 +19,11 @@ config ARCH_DEFCONFIG
 	string
 	default "arch/arm/defconfig"
 
+config CPU_HAS_FPU_DP
```
This option does not have a nomenclature that is specific to ARM. So, it should better be in the parent Kconfig file.

ioannisg reviewed (changes requested):
@aurel32 thanks for this PR. IMHO, however, the proposed solution is over-complicating things, with several new Kconfig option definitions. 

I wonder if we could just do the following:

- Create an Kconfig symbol: FP_DOUBLE_PRECISION, and define it inside: menu "Floating Point Options" in arch/Kconfig. This option could be selected by the SOC or Board that wants to use double precision. It should be selected along with CPU_HAS_FPU

- Clearly state CPU_HAS_FPU "alone" implies single precision, as @galak mentioned in #8374 

- in cmake, when selecting the compiler options, use CPU_HAS_FPU together with FP_DOUBLE_PRECISION to determine the right gcc option

aurel32 committed:
arch/Kconfig: add a hidden option to enable double precision FPU

This patch adds a hidden option to enable support for double-precision
FPU. It is selected by the architecture, CPU or SoC that has such an FPU.
This option is then used in fpu-for-gcc-m-cpu.cmake to select the
correct FPU.

Double precision FPU is enabled on x86 CPUs and SoCs, and on ARM on i.MX RT,
SAME70 SoCs and on STM32F746XG and STM32F769XI. This effectively disables
the double precision on STM32F723XE.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 commented:
I personally have the impression that the build system (i.e. cmake) is not the right place to hide processor properties, for instance the fact that a processor has a VFPv4 or VFPv5 FPU.

Anyway I have just pushed a new version of my PR which should match the implementation that you requested. From what I have found the ARC FPUs are all single precision, but I am not 100% sure about that.

ioannisg commented on arch/Kconfig:
```diff
@@ -241,7 +241,7 @@ config CPU_HAS_FPU
 	help
 	  This option is enabled when the CPU has hardware floating point
-	  unit.
+	  unit supporting single precision.
```
However: is this true for all architectures? (Looks OK for arm)

ioannisg replied:
Update: If the symbol FP_DOUBLE_PRECISION is only to be used for ARM compilation options, then I would propose, 
- CPU_HAS_FPU state that "single precision" is implied for ARM only,
- not to select FP_DOUBLE_PRECISION elsewhere, as it will have no effect
- and to get it more accurate, FP_DOUBLE_PRECISION could depend on ARM and also state at the description that it is needed for ARM CPUs to make this differentiation.

aurel32 replied:
At the moment it is indeed only needed on ARM, that's why the first version of my PR defined `CPU_HAS_FPU_DP` in `arch/arm/Kconfig`.

I'll modify the code so that it is an ARM only option.

ioannisg commented on arch/Kconfig:
```diff
@@ -269,6 +269,14 @@ config FLOAT
+config FP_DOUBLE_PRECISION
+	bool
+	# Omit prompt to signify "hidden" option
+	depends on FLOAT
```
I am not sure if this should depend on FLOAT. This is more of a hardware capability.

aurel32 replied:
It depends on `FLOAT` because it is in the `FLOAT`submenu, even if the option is hidden. `FLOAT` depends on 
`CPU_HAS_FPU`, so it indirectly depends on the hardware capability.

I can modify the code so that it also depends on `CPU_HAS_FPU`.

ioannisg commented on arch/Kconfig:
```diff
@@ -269,6 +269,14 @@ config FLOAT
+	# Omit prompt to signify "hidden" option
+	depends on FLOAT
+	help
+	  This option signifies the use of an FPU supporting double
```
Does it signify the "use"? Or, perhaps, the support? 
That said: could we reword the test to align with the description of CPU_HAS_FPU? But, no strong opinion on this .

aurel32 replied:
Ok, I'll try to reword the description.

ioannisg commented on arch/x86/Kconfig:
```diff
@@ -26,6 +26,7 @@ config CPU_ATOM
 	bool
 	select CMOV
 	select CPU_HAS_FPU
+	select FP_DOUBLE_PRECISION
```
I suppose, this mean that x86 has double precision always.
However: is this option actually used for x86?

I updated the comment of CPU_HAS_FPU reflecting my question, here.

erwango reviewed (commented):
Looks ok from my side, but please rebase.

aurel32 commented:
Closing this PR in favor of PR #13181
